docs/ha-considerations.md

```diff
@@ -0,0 +1,337 @@
+- `${HOST1_ID}` a symbolic name for the first load-balanced API Server host
+- `${HOST1_ADDRESS}` a resolvable address (DNS name, IP address) for the first load-balanced API Server host
+- additional `server` lines, one for each load-balanced API Server host
+
+### Option 1: Run the services on the operating system
+
+In order to run the two services on the operating system, the respective distribution's package manager can be used to install the software. This can make sense if they will be running on dedicated hosts not part of the Kubernetes cluster.
+
+On a RedHat-style Linux distribution this will most likely look like this:
+```
+# yum install keepalived haproxy
+```
+Having now installed the abovementioned configuration, the services can be enabled and started. On a recent RedHat-based system, `systemd` will be used for this:
+```
+# systemctl enable haproxy --now
+# systemctl enable keepalived --now
+```
+With the services up, now the Kubernetes cluster can be bootstrapped using `kubeadm init` (see [below](#bootstrap-the-cluster)).
+
+### Option 2: Run the services as static pods
+
+If `keepalived` and `haproxy` will be running on the control plane nodes they can be configured to run as static pods. All that is necessary here is placing respective manifest files in the _/etc/kubernetes/manifests_ directory before bootstrapping the cluster. During the bootstrap process, `kubelet` will bring the processes up, so that the cluster can use them while starting. This is an elegant solution, in particular with the setup described under [Stacked control plane and etcd nodes](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/#stacked-control-plane-and-etcd-nodes), where the `etcd` cluster, too, will be running as static pods on the control plane nodes.
+
+For this setup, two manifest files need to be created in _/etc/kubernetes/manifests_ (create the directory first).
+
+The manifest for `keepalived`, _/etc/kubernetes/manifests/keepalived.yaml_:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  creationTimestamp: null
+  name: keepalived
+  namespace: kube-system
+spec:
+  containers:
+  - image: osixia/keepalived:1.3.5-1
+    name: keepalived
+    resources: {}
+    securityContext:
+      capabilities:
+        add:
+        - NET_ADMIN
+        - NET_BROADCAST
+        - NET_RAW
+    volumeMounts:
+    - mountPath: /usr/local/etc/keepalived/keepalived.conf
+      name: config
+    - mountPath: /etc/keepalived/check_apiserver.sh
+      name: check
+  hostNetwork: true
+  volumes:
+  - hostPath:
+      path: /etc/keepalived/keepalived.conf
+    name: config
+  - hostPath:
+      path: /etc/keepalived/check_apiserver.sh
+    name: check
+status: {}
+```
+
+The manifest for `haproxy`, _/etc/kubernetes/manifests/haproxy.yaml_:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: haproxy
+  namespace: kube-system
+spec:
+  containers:
+  - image: haproxy:2.1.3-alpine
```

Comment:
wondering if alpine should not be recommended due to it's musl usage and related DNS issues.

Reply:
> wondering if alpine should not be recommended due to it's musl usage and related DNS issues.

Good point.